### PR TITLE
Optimize CRF Decoding with AVX/AVX2/AVX512F instruction

### PR DESCRIPTION
**Pull request by yihuaxu**
According to  the performance status of CRF Decoding, just implemented the intrinsic function's optimization to accelerate the data processing.  

**Platform:**  Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz /  Intel(R) Xeon(R) CPU E5-2699 v3 @ 2.30GHz
**Model Path:** [models/fluid/chinese_ner](https://github.com/PaddlePaddle/models/tree/develop/fluid/chinese_ner)
**Batch Size**: 6, 1, 12
**Command:** python infer.py --device CPU --profile
**Data Source:** Use original date provided in the model Chinese_NER in the github.

The following is the comparison with the different scenarios. 
l_ce40 tmp

### CODE REVIEW

**Comment by luotao1** [comment]
For code style, you can install pre-commit to auto modify the style. see https://github.com/PaddlePaddle/Paddle/blob/develop/CONTRIBUTING.md

**Comment by yihuaxu** [comment]
As the next step, there is one idea to create the "AVXDeviceContext", "AVX2DeviceContext" and "AVX512DeviceContext", likely "MKLDNNDeviceContext", to adapt the various CPU automatically via CPUID's checking. It should be one way to replace the Macros, like as "\_\_AVX\_\_", "\_\_AVX2\_\_" and etc,  to compile the hard code. It will be better than JIT code to balance the efficacy  and compatibility.

**Comment by tensor-tang** [comment]
It's cool to get such boost.

BTW May I ask how about the perf at batch size 1?

File: paddle/fluid/operators/crf_decoding_op.h [hunk 1]
```diff
@@ -85,6 +85,170 @@ class CRFDecodingOpKernel : public framework::OpKernel<T> {
...
+      size_t remain = tag_num % step_size;
+      int last_offset = static_cast<int>(remain) - static_cast<int>(step_size);
+
+      // Setup the alpha initial value.
+      size_t i_offset = 0;
+      for (size_t i = 0; i <= steps; ++i) {
+#ifdef __AVX512F__
+        __m512 w_content, x_content, alpha_content;
+
+        w_content = _mm512_loadu_ps((const float*)(w + i_offset));
```
Comment (luotao1) [inline paddle/fluid/operators/crf_decoding_op.h]: Could you add the explanation of some important`_mm512/256` lines? for example, what's the meaning of `_mm512_loadu_ps`? It's so hard for developers to read the code.

File: paddle/fluid/operators/crf_decoding_op.h [hunk 2]
```diff
@@ -85,6 +85,170 @@ class CRFDecodingOpKernel : public framework::OpKernel<T> {
     int* track_value =
         track.mutable_data<int>(emission_dims, platform::CPUPlace());
```
Comment (luotao1) [inline paddle/fluid/operators/crf_decoding_op.h:87]: Could you add the whole optimization method explanation of CRF decoding?

**Review by luotao1 — COMMENTED** [review]
please sign our Contributor License Agreement

File: paddle/fluid/operators/crf_decoding_op.h (hunk 1 above)
Comment (tensor-tang) [reply paddle/fluid/operators/crf_decoding_op.h]: I thought it's not necessary to add for every lines, give some key message should be fine.

File: paddle/fluid/operators/crf_decoding_op.h (hunk 2 above)
Comment (tensor-tang) [reply paddle/fluid/operators/crf_decoding_op.h:87]: Give some brief explanation should be fine.

File: paddle/fluid/operators/crf_decoding_op.h (hunk 1 above)
Comment (yihuaxu) [reply paddle/fluid/operators/crf_decoding_op.h]: The meaning of "_mm512_loadu_ps" is to load the content from un-aligned address.  These are some features of intrinsic function to optimize the data process.

File: paddle/fluid/operators/crf_decoding_op.h (hunk 2 above)
Comment (yihuaxu) [reply paddle/fluid/operators/crf_decoding_op.h:87]: The core of optimization concept is to use the AVX or AVX512 instruction to deal the float data as 8 or 16 vector every iteration.  and it need avoid the data transfer and other unnecessary operations.

**Review by luotao1 — APPROVED** [review]
LGTM！Excellent job!

**Pull request merged**